### PR TITLE
GH-542 Fixed table header and first column

### PR DESCRIPTION
**Pull request by dtsnell4**
Adds functionality for a fixed table header and fixed first column for the table widget.

Fixes GH-542

Signed-off-by: Dan Snell <dan.snell@lexisnexis.com>

### CODE REVIEW

**Comment by dtsnell4** [comment]
@GordonSmith @mlzummo @jbrundage 

Here is rawgit for demo: http://rawgit.com/dtsnell4/Visualization/Table_Widget_Fixed_Header_542/demos/dermatology.html?src/other/Table

I obviously multiplied the testData to be able to see the scrollbars.

Lint is failing due to not finding file "gulp-mocha-phantom.js".  Please advise.

**Comment by dtsnell4** [comment]
One other thing popped up this afternoon as I was putting together this PR.  @mlzummo received an email (and I believe @GordonSmith  did as well) from someone asking for this feature, but just for the header, not the first column.  I implemented it as either both or none.  Should I go back and separate the two, so you can have either-or?  Or would that be a future release?

**Comment by mlzummo** [comment]
yes i think you should

File: src/other/Table.js [hunk 1]
```diff
@@ -17,36 +17,69 @@
...
+                [37.665074, -122.384375, "green-dot.png", 4525, 423545, 24354, 54, 23552345, 2455, 245435],
+                [32.690680, -117.178540, "", 4525, 423545, 24354, 54, 2552345, 2455, 2405435],
+                [39.709455, -104.969859, "", 4525, 423545, 24354, 54, 2552345, 2455, 2495435],
+                [41.244123, -95.961610, "", 4525, 423545, 24354, 54, 2552345, 2455, 2454835],
+                [32.688980, -117.192040, "", 4525, 423545, 24354, 54, 2552345, 2455, 2475435],
+                [45.786490, -108.526600, "", 4525, 423545, 24354, 54, 2552345, 2455, 2456435],
+                [45.796180, -108.535652, "", 4525, 423545, 24354, 54, 2552345, 2455, 2455435],
+                [45.774320, -108.494370, "", 4525, 423545, 24354, 54, 2552345, 2455, 2445435],
+                [45.777062, -108.549835, "red-dot.png", 4525, 423545, 24354, 54, 25523345, 2455, 245435]
             ])
```
Comment (dtsnell4) [inline src/other/Table.js]: I was planning on removing the extra data once this code was finalized.  However, do we want to use a larger data set than before so developers can see the new features?

File: src/other/Table.js [hunk 2]
```diff
@@ -17,36 +17,69 @@
...
+            .fixedHeader(true)
         ;
         return this;
     };
 
-    Table.prototype.publish("pagination", false, "boolean", "enable or disable pagination",null,{tags:["Private"]});
+    Table.prototype.publish("pagination", false, "boolean", "Enable or disable pagination",null,{tags:['Private']});
+//    should the following default to true?
+    Table.prototype.publish("fixedHeader", false, "boolean", "Enable or disable fixed table header and first column",null,{tags:['Private']});
```
Comment (dtsnell4) [inline src/other/Table.js]: Do we want the table to default to fixed headers or normal mode?

File: src/other/Table.js [hunk 3]
```diff
@@ -17,36 +17,69 @@
...
     Table.prototype.publishProxy("itemsPerPage", "_paginator");
     Table.prototype.publishProxy("pageNumber", "_paginator", "pageNumber",1);
 
     Table.prototype.enter = function (domNode, element) {
         HTMLWidget.prototype.enter.apply(this, arguments);
-        this._parentElement.style("overflow", "auto");
 
-        this.table = element.append("table");
+        this.table = element.append("table").attr("id", "table" + this._id);
```
Comment (dtsnell4) [inline src/other/Table.js]: I need to remove the attribute "id" addition, my early implementation used it but no longer.

File: src/other/Table.js [hunk 4]
```diff
@@ -77,7 +110,9 @@
         return {
             _id: d,
             element: function () {
-                return context.tbody.selectAll("tr").filter(function (d2) { return d2 === d; });
+                return context.tbody.selectAll("tr").filter(function (d2) { 
+                    return d2 === d; 
+                });
             }
```
Comment (dtsnell4) [inline src/other/Table.js]: I can return this to its original state.  I was originally putting additional code here but removed it.

File: src/other/Table.js [hunk 5]
```diff
@@ -99,7 +134,11 @@
     Table.prototype.update = function (domNode, element) {
         HTMLWidget.prototype.update.apply(this, arguments);
         var context = this;
-
+        this._parentElement
+            .style("position", "relative")
+            .style("overflow", "hidden")
+        ;
+
```
Comment (dtsnell4) [inline src/other/Table.js]: This addition is no longer necessary and will be removed.

File: src/other/Table.js [hunk 6]
```diff
@@ -219,23 +272,190 @@
...
+            divrow.selectAll(".labels-wrapper").data([0])
+                .enter()
+                .append("table")
+                .classed("labels-wrapper", true)
+            ;
+            divrow.exit().remove();
+
+            var rowsWrapper = context._parentElement.select(".rows-wrapper");
+            var colsWrapper = context._parentElement.select(".cols-wrapper");
+            var colLabelsWrapper = colsWrapper.select('.labels-wrapper');
```
Comment (dtsnell4) [inline src/other/Table.js]: It turns out divcol = colsWrapper and divrow = rowsWrapper.  (it did not start that way) I will combine these.

File: src/other/Table.js [hunk 7]
```diff
@@ -219,23 +272,190 @@
...
+            }
+            function _copyLabelContents(){
+                colLabelsWrapper.html(theadSelection.html());
+                colLabelsWrapper
+                    .style("width", context.table.style("width"))
+                ;
+                
+                var origThead = element.selectAll("th");
+                var newThead = context._parentElement.selectAll(".cols-wrapper th");
+                origThead[0].forEach(function(el, i){
```
Comment (dtsnell4) [inline src/other/Table.js]: This should just use d3's .each() function.  Not sure why I did this.  Similar for following line

File: src/other/Table.js [hunk 8]
```diff
@@ -219,23 +272,190 @@
...
+                var borderWidth = parseInt(context.table.select("td").style("border-width"));
+                var rowSelection = context.table.selectAll('tbody > tr > td:first-child');
+                var rowWrapperWidth = rowSelection.node().getBoundingClientRect().width;
+                var theadWidth = parseInt(rowWrapperWidth) + parseInt(2 * borderWidth);
+                var rowContents = "<thead><tr>";
+                rowContents += "<th style='width:" + theadWidth + "px'>";
+                rowContents += origThead[0][0].innerHTML + "</th></tr></thead>";
+                
+                rowSelection[0].forEach(function(row){
+                    rowContents += '<tr><td class="row-label">'+ row.innerHTML +'</td></tr>';
```
Comment (dtsnell4) [inline src/other/Table.js]: Same as above

File: src/other/Table.js (hunk 1 above)
Comment (GordonSmith) [inline src/other/Table.js]: Its fine - we have another ticket which is a more general "move testData out of widgets" comment.  And one we can now do cleanly with the unit test framework in place.

File: src/other/Table.js [hunk 9]
```diff
@@ -125,11 +164,11 @@
             })
         ;
         th.select(".thIcon")
-            .text(function (column) {
+            .text(function (column, d) {
```
Comment (GordonSmith) [inline src/other/Table.js]: This might be confusing for some as we tend to use "d, idx" or "d, i" - this should be "column, idx"?

Comment (dtsnell4) [inline src/other/Table.js]: You are absolutely correct. I will change it.

**Comment by dtsnell4** [comment]
@GordonSmith Please re-review.  Also, do you have an opinion on separating the fixed header and the fixed first column into two distinct features?

**Comment by GordonSmith** [comment]
I think it should be two publish properties (but one pull request) - I look forward to playing with this.

**Comment by dtsnell4** [comment]
Ok, no problem.

**Comment by GordonSmith** [comment]
This is broken on:  http://rawgit.com/dtsnell4/Visualization/Table_Widget_Fixed_Header_542/demos/test.html?src/other/Table

**Comment by GordonSmith** [comment]
Closing in favor of:  https://github.com/hpcc-systems/Visualization/pull/756

**Pull request closed without merging**